Memory breakpoints

RipleyTom opened the pull request:
For hackers.
![image](https://user-images.githubusercontent.com/6031700/163454509-49c5c8ec-ff0e-412e-913c-d1d26b7b5ecb.png)
 
Edited, -DHAS_MEMORY_BREAKPOINTS=ON needs to be added for the option to be available within the UI.

Megamouse commented on rpcs3/Emu/Cell/PPUInterpreter.cpp:
```diff
@@ -25,6 +25,20 @@
+#endif
+	if (!status && ppu.check_state())
+	{
+		return;
```
is this your way of using the check_state return value?

elad335 replied:
this is a typo.

RipleyTom replied:
No idea if it is necessary but I'm pretty sure I called check_state to make sure the state change was handled immediately.
I changed the code to make it clearer.

Megamouse commented on rpcs3/Emu/Cell/PPUInterpreter.cpp:
```diff
@@ -25,6 +25,20 @@
+#endif
+	if (!status && ppu.check_state())
+	{
+		return;
```
is this your way of using the check_state return value?

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.h:
```diff
@@ -3,10 +3,17 @@
 #include "util/types.hpp"
 
 #include <QListWidget>
+#include <QDialog>
+#include <QVBoxLayout>
+#include <QLineEdit>
+#include <QLabel>
+#include <QComboBox>
+#include <QPushButton>
```
why all the includes?

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -65,19 +66,34 @@ void breakpoint_list::RemoveBreakpoint(u32 addr)
-
-	QString text = qstr(m_disasm->last_opcode);
-	text.remove(10, 13);
+	QString breakpointItemText;
```
snake_case

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+
+void breakpoint_list::ShowAddBreakpointWindow()
+{
+	QDialog* diag = new QDialog(this);
```
what's a diag? Just write dialog or dlg

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+	QHBoxLayout* hbox_top = new QHBoxLayout();
+	QLabel* l_address     = new QLabel(tr("Address"));
+	QLineEdit* t_address  = new QLineEdit();
+	t_address->setPlaceholderText("Address here");
```
```suggestion
	t_address->setPlaceholderText(tr("Address here"));
```

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+	QHBoxLayout* hbox_bot = new QHBoxLayout();
+	QComboBox* co_bptype  = new QComboBox(this);
+	QStringList qstr_breakpoint_types;
+	qstr_breakpoint_types << "Memory Read"
+						  << "Memory Write"
+						  << "Memory Read&Write"
+						  << "Execution";
```
weird indentations, also missing tr()

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+{
+	QDialog* diag = new QDialog(this);
+
+	diag->setWindowTitle(tr("Add a breakpoint"));
```
dlg->setAttribute(Qt::WA_DeleteOnClose);

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+		}
+	}
+
+	diag->deleteLater();
```
remove this, see above

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+			}
+
+			if (bp_t)
+				AddBreakpoint(address, bp_t);
```
i think this whole switch would be simpler if you just directly called AddBreakpoint

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -176,3 +196,82 @@ void breakpoint_list::OnBreakpointListDelete()
+	vbox_panel->addLayout(hbox_bot);
+
+	QHBoxLayout* hbox_buttons = new QHBoxLayout();
+	QPushButton* b_cancel     = new QPushButton(tr("Cancel"));
+	QPushButton* b_addbp      = new QPushButton(tr("Add"));
+
+	hbox_buttons->addWidget(b_cancel);
+	hbox_buttons->addWidget(b_addbp);
+	vbox_panel->addLayout(hbox_buttons);
+
+	diag->setLayout(vbox_panel);
+
+	connect(b_cancel, &QAbstractButton::clicked, diag, &QDialog::reject);
+	connect(b_addbp, &QAbstractButton::clicked, diag, &QDialog::accept);
```
QDialogButtonBox* button_box = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
	button_box->button(QDialogButtonBox::Ok)->setText(tr("Add"));
	connect(button_box, &QDialogButtonBox::accepted, this, &QDialog::accept);
	connect(button_box, &QDialogButtonBox::rejected, this, &QDialog::reject);

Megamouse commented on rpcs3/Emu/Memory/vm.h:
```diff
@@ -306,29 +321,47 @@ namespace vm
 		{
 			_ref<u64>(addr) = value;
+
+			if (ppu && g_breakpoint_handler->HasBreakpoint(addr, breakpoint_types::bp_write))
```
g_breakpoint_handler can be null in all cases in this file

Megamouse commented on rpcs3/rpcs3qt/breakpoint_handler.h:
```diff
@@ -1,13 +1,15 @@
 	bp_read = 0x1,
 	bp_write = 0x2,
 	bp_exec = 0x4,
+	__bitset_enum_max
```
unused?

RipleyTom replied:
Necessary to store the value in a bs_t.

nastys commented:
Putting invalid characters e.g. ``OX162CDB3`` instead of ``0x162CDB3`` in the input box for the BPM address causes RPCS3 to crash.

illusion0001 commented:
Adding execution breakpoints while game is running crashes

RipleyTom commented:
Updated and cleaned the PR and it is now meant for merging.
Note that cmake needs to be called with -DHAS_MEMORY_BREAKPOINTS=ON for the option to be available.

Megamouse commented on rpcs3/Emu/Cell/PPUInterpreter.cpp:
```diff
@@ -26,6 +26,34 @@
+{
+	if (!g_breakpoint_handler.IsBreakOnBPM()) return;
+
+	bool status = ppu.state.test_and_set(cpu_flag::dbg_pause);
```
```suggestion
	const bool status = ppu.state.test_and_set(cpu_flag::dbg_pause);
```

Megamouse commented on rpcs3/Emu/Memory/vm.h:
```diff
@@ -276,29 +299,65 @@ namespace vm
+#ifdef RPCS3_HAS_MEMORY_BREAKPOINTS
+			if (ppu && g_breakpoint_handler.HasBreakpoint(addr, breakpoint_types::bp_write))
+			{
+				debugbp_log.success("BPMW: breakpoint writing(1) 0x%x at 0x%x", value, addr);
```
Why 1 and not 64?

Megamouse commented on rpcs3/rpcs3.vcxproj:
```diff
@@ -960,6 +967,16 @@
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">$(QTDIR)\bin\moc.exe;%(FullPath);$(QTDIR)\bin\moc.exe;%(FullPath)</AdditionalInputs>
       <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Release|x64'">$(QTDIR)\bin\moc.exe;%(FullPath);$(QTDIR)\bin\moc.exe;%(FullPath)</AdditionalInputs>
     </CustomBuild>
+    <CustomBuild Include="rpcs3qt\debugger_add_bp_window.h">
```
I don't know if anything changed here since the PR was maybe.
Maybe it's worth double checking if all the params are up to date

RipleyTom replied:
That part is new so it should be fine, I just copy/pasted entry below and changed the appropriate stuff.

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -72,19 +73,34 @@ void breakpoint_list::RemoveBreakpoint(u32 addr)
-	text.remove(10, 13);
+	QString breakpoint_item_text;
+	if (type == breakpoint_types::bp_exec)
+	{
```
Would a switch case work here?

RipleyTom replied:
No, switch won't work with bitsets.

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -194,13 +210,8 @@ void breakpoint_list::OnBreakpointListDoubleClicked()
+void breakpoint_list::OnBreakpointListRightClicked(const QPoint& pos)
 {
-	if (!itemAt(pos))
-	{
```
Why was this removed again?
Don't remember the logic here...

elad335 commented on rpcs3/Emu/Cell/PPUInterpreter.cpp:
```diff
@@ -26,6 +26,34 @@
+
+	bool status = ppu.state.test_and_set(cpu_flag::dbg_pause);
+#ifdef WITH_GDB_DEBUGGER
+	fxm::get<GDBDebugServer>()->notify();
```
This won't compile, we use FXO now.

RipleyTom replied:
I removed that whole ifdef, I couldn't find traces of GDBDebugServer or WITH_GDB_DEBUGGER in the rest of the source code.

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -215,7 +227,28 @@ void breakpoint_list::OnBreakpointListRightClicked(const QPoint &pos)
+	}
+
+	QAction* m_addbp = new QAction(tr("Add Breakpoint"), this);
+	connect(m_addbp, &QAction::triggered, this, [=, this]
```
```suggestion
	connect(m_addbp, &QAction::triggered, this, [this]
```

Megamouse commented on rpcs3/rpcs3qt/breakpoint_list.cpp:
```diff
@@ -215,7 +227,28 @@ void breakpoint_list::OnBreakpointListRightClicked(const QPoint &pos)
+
+#ifdef RPCS3_HAS_MEMORY_BREAKPOINTS
+	QAction* m_tglbpmbreak = new QAction(m_ppu_breakpoint_handler->IsBreakOnBPM() ? tr("Disable BPM") : tr("Enable BPM"), this);
+	connect(m_tglbpmbreak, &QAction::triggered, [&]
```
```suggestion
	connect(m_tglbpmbreak, &QAction::triggered, [this]
```

Megamouse commented on rpcs3/rpcs3qt/debugger_add_bp_window.cpp:
```diff
@@ -0,0 +1,115 @@
+				bp_t = breakpoint_types::bp_exec;
+				break;
+			default:
+				bp_t = {};
```
obsolete

Megamouse commented on rpcs3/rpcs3qt/debugger_add_bp_window.cpp:
```diff
@@ -0,0 +1,115 @@
+				return;
+			}
+
+			u32 address = static_cast<u32>(parsed_address);
```
```suggestion
			const u32 address = static_cast<u32>(parsed_address);
```

Megamouse commented on rpcs3/rpcs3qt/debugger_add_bp_window.cpp:
```diff
@@ -0,0 +1,115 @@
+#include <QDialogButtonBox>
+#include <QMessageBox>
+
+debugger_add_bp_window::debugger_add_bp_window(breakpoint_list* bp_list, QWidget* parent)
```
ensure bp_list ?

VelocityRa commented:
Unfortunately this doesn't work (for writes) for several instructions: `STVX`, `STVXL`, `STDBRX`, `STWBRX`, `STFSX`, `STFSUX`, `STFDX`, `STFDUX`, `STHBRX`, `STFS`, `STFSU`, `STFD`, `STFDU`.

Because they write through `vm::_ref` instead of the `vm::write` methods.
